File: orm-reference-implementations/sql-alchemy/entity/student_card.py
from base import Base
from entity.student_card_id import StudentCardId
from sqlalchemy import String
from sqlalchemy.orm import Mapped, composite, mapped_column, relationship


class StudentCard(Base):
    __tablename__ = 'student_card'

    printed_name: Mapped[str] = mapped_column(String(255))

    card_nr: Mapped[str] = mapped_column("card_nr", String(255), primary_key=True)
    card_version: Mapped[str] = mapped_column("card_version", String(255), primary_key=True)
    id: Mapped[StudentCardId] = composite("card_nr", "card_version")
    study_programs: Mapped[list["StudentCardStudyProgram"]] = relationship(back_populates="student_card")
    student: Mapped["Student"] = relationship(back_populates="student_card")

File: orm-reference-implementations/entity-framework/entity/StudentCard.cs
using Microsoft.EntityFrameworkCore;
using System.ComponentModel.DataAnnotations.Schema;

namespace university.entity
{
    [Table("student_card")]
    [PrimaryKey(nameof(CardNr), nameof(CardVersion))]
    public class StudentCard
    {
        // LIMITATION: Owned types cannot be used in Table-per-Class inheritance hierarchies, embeddables must be mapped as flattened properties: https://github.com/dotnet/efcore/issues/32028
        [Column(TypeName = "Varchar(255)")]
        public string CardNr { get; set; }

        [Column(TypeName = "Varchar(255)")]
        public string CardVersion { get; set; }

        [Column(TypeName = "Varchar(100)")]
        public required string PrintedName { get; set; }

        public Student? Student { get; set; }

        public List<StudentCardStudyProgram>? StudyPrograms { get; set; }

    }
}
